### PR TITLE
Connect to wss without wsproxy

### PR DESCRIPTION
**Pull request by petuhovskiy**
Needed to work on https://github.com/neondatabase/neon/issues/3225

### CODE REVIEW

File: shims/net/index.ts [hunk 1]
```diff
@@ -157,8 +157,8 @@ export class Socket extends EventEmitter {
     this.ws = await new Promise<WebSocket>(resolve => {
       try {
         // ordinary/browser path
-        const wsProtocol = this.useSecureWebSocket ? 'wss:' : 'ws:';
```
Comment (jawj) [inline shims/net/index.ts]: Can we keep this conditional?

File: shims/net/index.ts [hunk 2]
```diff
@@ -157,8 +157,8 @@ export class Socket extends EventEmitter {
...
-        const ws = new WebSocket(wsProtocol + '//' + wsAddr);
+        const wsProtocol = 'wss:';
+        const ws = new WebSocket(wsProtocol + '//' + host + ':8443');
```
Comment (jawj) [inline shims/net/index.ts:161]: I'm guessing this is temporary, and in production we'll listen on 443?

**Review by jawj — CHANGES_REQUESTED** [review]
Can we not make the `useSecureWebSocket` option a no-op? Otherwise this looks good to merge into main.

File: shims/net/index.ts (hunk 1 above)
Comment (petuhovskiy) [reply shims/net/index.ts]: Yes, we can allow using `ws:` in the driver, but I'm not sure it's a good idea to support unencrypted ws in the proxy itself. Note that this PR is more for experiments, rather than ready to merge.

Comment (jawj) [reply shims/net/index.ts]: Agreed, we should support wss: only. But others may want to use ws: (coupled with the experimental JS encryption that's also built into the driver).

File: shims/net/index.ts (hunk 2 above)
Comment (petuhovskiy) [reply shims/net/index.ts]: Yes, but it's more convenient to use `:8443` for debug and development. I changed the logic around `this.wsProxy` a bit, so it's possible to use custom port now.

**Comment by petuhovskiy** [comment]
I don't think it's ready to merge to main yet, but it can be merged to `no-webassembly` branch, and you can continue your work there.

What should be fixed before merging to main:
- our neon proxy should be deployed to proxy, currently it's not even on staging yet
- build should be ran without debug
- probably additional work should be done to improve backwards compatibility (currently this PR has breaking changes), but not sure if it worth it

In other words, this code is not yet ready to use and shouldn't be in main yet. WDYT @jawj?

**Comment by jawj** [comment]
@petuhovskiy Sure, it's fine to stick to the `no-webassembly` branch (and we certainly shouldn't publish to npm yet).

How soon do you think we could deploy to staging?

**Comment by petuhovskiy** [comment]
I want to get a proxy review from @funbringer, probably make some fixes and then we can merge and deploy it. I think it should get to staging around next week.

**Comment by jawj** [comment]
OK. We've more or less promised Vercel to launch next week, so it would be great if we could deploy to staging early in the week (and to production by the end of the week?).

**Review by jawj — APPROVED** [review]

**Comment by kelvich** [comment]
wss proxy is on staging now. @petuhovskiy can you rebase this?

**Comment by petuhovskiy** [comment]
Sorry, forgot to update this PR. Will rebase it now.

**Comment by petuhovskiy** [comment]
Updated the PR, I used `postgres://arthur:pXFNr4kVui0Z@ep-patient-mode-143675.us-east-2.aws.neon.build/neondb` (staging db) in all variables.

**Comment by nicksrandall** [comment]
Is it possible yet for me to test this? I'm working on a POC and want to get a ballpark measurement on what latency is like with the new proxy.

**Comment by kelvich** [comment]
@nicksrandall we are still doing some work to reduce the latency (mainly around making authentication quicker in happy path https://github.com/neondatabase/neon/pull/3331), hence no public release yet. However, things from this PR are already implemented in `no-webassembly` branch and published in temporary `@jawj/test-serverless` npm package. Proxy with websockets is on prod in all regions, so feel free to test it. `@jawj/test-serverless` will switch to websockets if it detects `neon.tech` in connection string. Here is an example: https://github.com/jawj/edge-data-latency-mine/blob/main/pages/api/neon-global.ts#L13-L23

Curious to see your results.

(This PR is ought to be closed, but i'll keep it open for now to keep this discussion)

**Comment by jawj** [comment]
OK, equivalent changes have now been made, merged in, and released to npm as a new version of `@neondatabase/serverless`, therefore closing.

https://community.neon.tech/ may be useful for further discussion.
